### PR TITLE
Better VIMWIKI_MARKDOWN_EXTENSIONS environment variable parsing

### LINKED ISSUE #16: Support For linenums for CodeHilite
line number can be enabled by  passing [linenums](https://python-markdown.github.io/extensions/code_hilite/#usage) to codehilite

```python
Codehilite(linenums=True)
```

If i want to raise PR which approach is advised.

1. create `VIMWIKIMARKDOWN_LINENUMS` for passing this arguement
2. parse `VIMWIKIMARKDOWN_EXTENSION` using `ast.literal_eval` and then add default plugins if they are not available

### PR DESCRIPTION
**Pull request by sak96**
1. Json list syntax of extension. `["toc", "nl2br"]`.
1. Json dictionary syntax of extension with configuration
	 `{"toc": {"baselevel": 2 }, "nl2br": {}}`.
	 **Note**: `{}` configuration implies no configuration.
1. [LEGACY] comma separated list of extensions `toc,nl2br`.

### CODE REVIEW

**Comment by sak96** [comment]
@WnP  this is more configurable. please do check this out.

**Comment by sak96** [comment]
@WnP  please review.

File: vimwiki_markdown.py [hunk 1]
```diff
@@ -93,12 +94,36 @@ def main():
     filename, _ = os.path.splitext(os.path.basename(INPUT_FILE))
     output_file = os.path.join(OUTPUT_DIR, filename + ".html")
 
-    extensions = ["fenced_code", "tables"]
-    extensions += os.getenv("VIMWIKI_MARKDOWN_EXTENSIONS", "").split(",")
-    extensions = set([e for e in extensions if e] + [CodeHiliteExtension()])
+    # parse extension into dict
+    try:
```
Comment (WnP) [inline vimwiki_markdown.py:102]: This try block should only contain `ast` call

File: vimwiki_markdown.py [hunk 2]
```diff
@@ -93,12 +94,36 @@ def main():
...
+        if isinstance(extensions_env, list):
+            extensions = extensions_env
+            extension_configs = {}
+        elif isinstance(extensions_env, dict):
+            extensions = list(extensions_env.keys())
+            extension_configs = extensions_env
+        else:
+            extensions = default_extension
+            extension_configs = {}
+    except (ValueError, SyntaxError):
```
Comment (WnP) [inline vimwiki_markdown.py]: this except block should only set `extensions_env`

File: vimwiki_markdown.py [hunk 3]
```diff
@@ -93,12 +94,36 @@ def main():
...
+            os.getenv(
+                "VIMWIKI_MARKDOWN_EXTENSIONS",
+                "{}",
+            )
+        )
+
+        # extract extension and extension_configs
+        if isinstance(extensions_env, list):
+            extensions = extensions_env
+            extension_configs = {}
```
Comment (WnP) [inline vimwiki_markdown.py]: What if I'd like to have multiple extensions and set some options for each one (or few ones)?

**Review by WnP — COMMENTED** [review]
I like this solution, at it is backward compatible.

But I also like to be able to set something like:

```viml
let $VIMWIKI_MARKDOWN_EXTENSIONS="[{'codehilite': {'linenums': True}}, 'fenced_code', 'tables']"
```

but it's no possible at the moment.

I see how we can can change the code to handle that (a simple for loop may do the trick)

Let me know if you want to do it, otherwise I'll (whatever happen I'll merge this PR to preserve you name in contributor list)

Thanks for contributing to vim wiki mardown :-)

**Comment by sak96** [comment]
```vim
let $VIMWIKI_MARKDOWN_EXTENSIONS="{'codehilite': {'linenums': True}, 'fenced_code': {}, 'tables' : {}}"
```
@WnP updated readme to explain the same.

---
I can do it in separate PR.

```diff
     if isinstance(extensions_env, list):
         extensions = extensions_env
-        extension_configs = {}
+        for extension in extensions_env:
+            if isinstance(extension, dict):
+                extension_configs.update(extension)
+            elif isinstance(extension, str):
+                extensions.append(extension)
+            # TODO: what do we do here ??
+        # TODO: is this really required ??
+        extensions.extend(extension_configs.keys())
     elif isinstance(extensions_env, dict):
         extensions = list(extensions_env.keys())
```

File: vimwiki_markdown.py (hunk 3 above)
Comment (sak96) [reply vimwiki_markdown.py]:
```vim
let $VIMWIKI_MARKDOWN_EXTENSIONS="{'codehilite': {'linenums': True}, 'fenced_code': {}, 'tables' : {}}"
```

**Comment by WnP** [comment]
hi @sak96,

In fact I see it way simpler, something like the following should be enough:

```py
import json

# parse extensions
extensions_env = os.getenv("VIMWIKI_MARKDOWN_EXTENSIONS", "{}")
try:
    extensions = json.loads(extensions_env)
except json.decoder.JSONDecodeError:
    extensions = extensions_env.split(',')

if isinstance(extensions, list):
    extensions = {e: {} for e in extensions}

# Setup markdown parser
md = markdown.Markdown(
    extensions=set(
        list(extensions.keys()) + ["fenced_code", "tables", "codehilite"]
    ),
    extension_configs=extensions,
)
```

IMHO, `ast.iteral_eval` is overkill and not safe. `json` sounds more appropriate here.

The colon separated list format should be deprecated and will be removed in future release.

No need for a separate pull request. Also please squash your commits.

Let me know your thoughts.

Thanks

**Review by WnP — COMMENTED** [review]
see my previous comment

**Comment by sak96** [comment]
done

**Review by WnP — COMMENTED** [review]
Thanks @sak96, that's perfect.

Can you squash all your commits in one now?

**Comment by sak96** [comment]
done.

**Comment by sak96** [comment]
need to recheck

**Comment by sak96** [comment]
json load vs literal eval is
- literal eval can parse single quoted strings
- literal eval use python syntax (True instead of true of json)
...

TLDR: "json syntax" != python_syntax

**Comment by WnP** [comment]
> literal eval can parse single quoted strings

Yes, but look:

```viml
let $VIMWIKI_MARKDOWN_EXTENSIONS='{"codehilite": {"linenums": true}, "fenced_code": {}, "tables" : {}}'
```
vs

```viml
let $VIMWIKI_MARKDOWN_EXTENSIONS="{'codehilite': {'linenums': True}, 'fenced_code': {}, 'tables' : {}}"
```

... can't see where the issue is.

> literal eval use python syntax (True instead of true of json)

Here we just need to inject some data / parameter / config, we do not need to eval whatever python code. That's why json is more appropriate.
